plugins/datastreamer_plugin/fastdds/Participant.cpp

```diff
@@ -0,0 +1,456 @@
+// Copyright 2022 Proyectos y Sistemas de Mantenimiento SL (eProsima).
+//
+// This file is part of eProsima Fast DDS Monitor.
+//
+// eProsima Fast DDS Monitor is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// eProsima Fast DDS Monitor is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with eProsima Fast DDS Monitor. If not, see <https://www.gnu.org/licenses/>.
+
+/**
+ * @file Participant.hpp
+ */
+
+#include <fastdds/dds/domain/DomainParticipantFactory.hpp>
+#include <fastrtps/xmlparser/XMLProfileManager.h>
+#include <fastrtps/types/DynamicDataHelper.hpp>
+#include <fastrtps/types/DynamicDataFactory.h>
+#include <fastrtps/types/TypeObjectFactory.h>
+
+#include "Participant.hpp"
+#include "utils/utils.hpp"
+
+namespace eprosima {
+namespace plotjuggler {
+namespace fastdds {
+
+////////////////////////////////////////////////////
+// READERHANDLER DELETER
+////////////////////////////////////////////////////
+
+ReaderHandlerDeleter::ReaderHandlerDeleter(
+        eprosima::fastdds::dds::DomainParticipant* participant,
+        eprosima::fastdds::dds::Subscriber* subscriber)
+    : participant_(participant)
+    , subscriber_(subscriber)
+{
+}
+
+void ReaderHandlerDeleter::operator ()(
+        ReaderHandler* reader) const
+{
+    // Stop data gathering
+    reader->stop();
+    // Delete reader
+    subscriber_->delete_datareader(reader->reader_);
+    // Delete topic
+    participant_->delete_topic(reader->topic_);
+}
+
+////////////////////////////////////////////////////
+// CREATION & DESTRUCTION
+////////////////////////////////////////////////////
+
+Participant::Participant(
+        eprosima::fastdds::dds::DomainId_t domain_id,
+        std::shared_ptr<TopicDataBase> discovery_database,
+        FastDdsListener* listener)
+    : listener_(listener)
+    , discovery_database_(discovery_database)
+{
+    // TODO check entities are created correctly
+
+    // Create Domain Participant
+    participant_ = eprosima::fastdds::dds::DomainParticipantFactory::get_instance()->create_participant(
+        domain_id,
+        default_participant_qos_(),
+        this,
+        default_listener_mask_());
```

Comment:
Safer and simpler to wrap a participant_ as a `std::unique_ptr` with a deleter that does `participant_->delete_subscriber(subscriber_);`

Reply:
Here has 3 main advantages:
* Safety: automatic custom resource release in case of exception throw (RAII https://en.cppreference.com/w/cpp/language/raii) (if release were not custom, a plain data member object would be fine).
* Maintenance: No need to remember to release the resource on edge cases that may arise in the future.
* Code readability / clarity / less cognitive load for the reader: declare initialization / finalization of resource within the same scope (just at the resource creation), not at two separate scopes (the constructor and destructor of its parent class, which may be arbitrarily separated)

Reply:
No need to worry about that, the same C++ standard defines that ordering: https://isocpp.org/wiki/faq/dtors#order-dtors-for-locals
So the code is safer and simpler - logic in the Participant destructor is unnecessary (again, think twice if you think you need a custom destructor in modern C++). If you think this does not worth the effort nevermind, go ahead with the merge